### PR TITLE
pip: Generate single module with full sources

### LINKED ISSUE #157: flatpak-pip-generator: unexpected behaviour for packages that rely on wheel
The flatpak-pip-generator produces incomplete json modules for packages that rely on wheel (not packaged as wheels, but list it as dependency).

**Steps to reproduce:**
Run
```
python flatpak-pip-generator.py debugpy
```
this outputs a file like this:
```json
{
    "name": "python3-debugpy",
    "buildsystem": "simple",
    "build-commands": [
        "pip3 install --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"debugpy\""
    ],
    "sources": [
        {
            "type": "file",
            "url": "https://files.pythonhosted.org/packages/92/78/6f8857235b6b67194af3e66d0767d102e32a3e4055d0212f1783c0b4e5b0/debugpy-1.1.0.zip",
            "sha256": "955af25acd01ddd6a7adf6e495b118ed2adf072c61294a7e899c2c32f64e9a60"
        }
    ]
}
```
Adding this in a flatpak manifest fails with "ERROR: Could not find a version that satisfies the requirement wheel". Adding a module with "wheel" in front of it does not help. The only thing that works is adding "wheel" and "setuptools" sources to the module like this:
```json
{
      "name": "python3-debugpy",
      "buildsystem": "simple",
      "build-commands": [
        "pip3 install --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST}  \"debugpy\""
      ],
      "sources": [
        {
          "type": "file",
          "url": "https://files.pythonhosted.org/packages/92/78/6f8857235b6b67194af3e66d0767d102e32a3e4055d0212f1783c0b4e5b0/debugpy-1.1.0.zip",
          "sha256": "955af25acd01ddd6a7adf6e495b118ed2adf072c61294a7e899c2c32f64e9a60"
        },
        {
          "type": "file",
          "url": "https://files.pythonhosted.org/packages/a7/00/3df031b3ecd5444d572141321537080b40c1c25e1caa3d86cdd12e5e919c/wheel-0.35.1-py2.py3-none-any.whl",
          "sha256": "497add53525d16c173c2c1c733b8f655510e909ea78cc0e29d374243544b77a2"
        },
        {
          "type": "file",
          "url": "https://files.pythonhosted.org/packages/6d/38/c21ef5034684ffc0412deefbb07d66678332290c14bb5269c85145fbd55e/setuptools-50.3.2-py3-none-any.whl",
          "sha256": "2c242a0856fbad7efbe560df4a7add9324f340cf48df43651e9604924466794a"
        }
      ]
    }
```

### PR DESCRIPTION
**Pull request by nanonyme**
This should work correctly with build isolation and fix #157
Essentially we generate a single module with a lot of sources, build wheels and finally install said wheels.

### CODE REVIEW

File: pip/flatpak-pip-generator [hunk 1]
```diff
@@ -275,6 +260,7 @@ with tempfile.TemporaryDirectory(prefix=tempdir_prefix) as tempdir:
             source = OrderedDict([
                 ('type', vcs),
                 ('url', url),
+                ('subdir', name),
```
Comment (gasinvein) [inline pip/flatpak-pip-generator:263]: I guess it's `dest`, not `subdir`?

Comment (nanonyme) [reply pip/flatpak-pip-generator:263]: Changes

**Comment by nanonyme** [comment]
I just found out there's something odd happening when this is ran with --runtime so closing this for now so no one accidentally merges it.

**Comment by gasinvein** [comment]
@leinardi @tfmoraes Can you please check how this it works for you?

**Comment by leinardi** [comment]
> @leinardi @tfmoraes Can you please check how this it works for you?

Hi, sorry if I ask something stupid but, what should I test exactly? Should I just check out this branch and run `flatpak-pip-generator` with my default parameters (`--requirements-file=../requirements.txt --output pypi-dependencies`)?

BTW, today I tried to run `flatpak-pip-generator` form master and it seems to be broken for me:

1. the file `flatpak-pip-generator` lost the execution permission: now I have to manually chmod it to make executable
2. when I run it I get this `SyntaxError`:
```
$ ./flatpak-builder-tools/pip/flatpak-pip-generator --requirements-file=../requirements.txt --output pypi-dependencies
  File "./flatpak-builder-tools/pip/flatpak-pip-generator", line 34
    help='https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support')
    ^
SyntaxError: invalid syntax
```
These issues seems related to the changes made on the 4th of December. The version 167e028582d744fe5721de71baa5ad7ce9e3719e still works fine.
Should I open a separate bug for this?

**Comment by gasinvein** [comment]
@leinardi yes, the generator on master is mostly broken currently. This PR should fix (at least to some extent).
>Should I open a separate bug for this?

#167 was supposed to fix the missing comma.

**Comment by leinardi** [comment]
I've tested your branch `nanonyme/support-build-isolation` (cd8d46b381cf2a8677942383e0044a550a82d7dd):

1. flatpak-pip-generator is still missing the executable permission: could you please add it?
2. the execution of `flatpak-pip-generator` was successful (exit code 0)
3. The flatpak build fails with this error:
```
Processing ./pyusb-1.1.0.tar.gz
  Installing build dependencies ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /usr/lib/python3.8/site-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-ahpx1416/overlay --no-warn-script-location --no-binary :none: --only-binary :none: --no-index --find-links file:///run/build/pypi-dependencies -- 'setuptools>=30.3.0' wheel setuptools_scm
       cwd: None
  Complete output (4 lines):
  Looking in links: file:///run/build/pypi-dependencies
  Processing ./setuptools-51.0.0-py3-none-any.whl
  ERROR: Could not find a version that satisfies the requirement wheel (from versions: none)
  ERROR: No matching distribution found for wheel
  ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 /usr/lib/python3.8/site-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-ahpx1416/overlay --no-warn-script-location --no-binary :none: --only-binary :none: --no-index --find-links file:///run/build/pypi-dependencies -- 'setuptools>=30.3.0' wheel setuptools_scm Check the logs for full command output.
```

This is the `requirements.txt` used:
```
injector==0.18.4
hidapi==0.9.0.post3
liquidctl==1.3.3
matplotlib==3.2.2
peewee==3.13.3
PyGObject==3.38.0
pyxdg==0.27
requests==2.25.0
Rx==3.1.1
```

**Comment by gasinvein** [comment]
@leinardi That's something, thanks for the feedback.
@nanonyme Looks like the setuptools/wheel issue is the here. Do you have an idea where it comes from?

**Comment by nanonyme** [comment]
Nope. Thanks for failing example. Will do some further testing.

**Comment by leinardi** [comment]
@nanonyme for `hidapi` at some point you'll need to manually add this patch: https://github.com/flathub/com.leinardi.gkraken/blob/master/hidapi_fix_prefix.patch

With the old `flatpak-pip-generator` version I added it like this: https://github.com/flathub/com.leinardi.gkraken/blob/a46f6e2665845b8803272f6d862168ec257a0bc8/pypi-dependencies.json#L39

**Comment by gasinvein** [comment]
I guess the dependency chain here is `hidapi` -> `pyusb` -> `setuptools_scm` -> `setuptools` -> `wheel`?

**Comment by nanonyme** [comment]
I see. Well, that is definitely an argument against my change. It makes patching more awkward.

**Comment by nanonyme** [comment]
The problem here is setup_depends is not followed at all. If wheel is explicitly downloaded, then things work a bit better.

**Comment by tfmoraes** [comment]
@gasinvein apparently it generates the json file correctly. But unfortunately I can't test it because the CMake in SDK has this [bug](https://gitlab.kitware.com/cmake/cmake/-/issues/21529) that impedes GDCM to be compiled. I removed GDCM just to test the compilation, but InVesalius will not run.

**Comment by gasinvein** [comment]
@tfmoraes You generate modules one-by-one, right? Can you please also try generating all at once (probably from `requirements.txt` or by listing all direct dependencies in one cli run)?

**Comment by tfmoraes** [comment]
Nope, I generate all at once:

```
python flatpak-pip-generator wxpython==4.0.7.post2 numpy scipy imageio scikit-image Pillow pyserial psutil nibabel configparser h5py==2.10.0 PyPubsub plaidml-keras cython
```

**Comment by gasinvein** [comment]
@tfmoraes And the generated `python3-modules.json` builds fine?

**Comment by tfmoraes** [comment]
@gasinvein It just finished the compilation and it worked. It just needed to add an `env` variable because of wxPython. The `python3-modules.json` generated is this: https://gist.github.com/tfmoraes/4ab7ff39f2b231e4e259e04c1277cf06

Because of that bug CMake with GDCM I can't test if InVesalius is running correctly. Generate one file with all python modules eases a lot the process of packaging.

**Comment by tfmoraes** [comment]
I submitted a [bug report](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1180) to the Freedesktop-Sdk about this problem with CMake. After this being fixed I'll test it correctly. I entered inside the InVesalius Flatpak using this command:

```
flatpak run --command=python br.gov.cti.invesalius
```
And started to import the python modules and it imported all python modules that I tested.

**Comment by nanonyme** [comment]
@leinardi could you please also test whether https://github.com/flatpak/flatpak-builder-tools/pull/169 works better or worse for you?

**Comment by nanonyme** [comment]
In case someone wants to give another try at this, feel free to look into diff in this PR. I'm not intending to keep working on this.

**Pull request closed without merging**